### PR TITLE
resumable iteration

### PR DESCRIPTION
**Pull request by jmtroth0**
adds both parts of resumability:
1) the main iterators use the `ResumableIterator`s classes
2) the queues store a persistent list of ids in redis that are currently processing or are in the queues so that they can be rebuilt

It uses timestamp to control runs, which is printed out at the beginning of the run

@snopoke @millerdev @emord

### CODE REVIEW

File: corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py [hunk 1]
```diff
@@ -48,6 +48,7 @@ def add_arguments(self, parser):
         parser.add_argument('--show-diffs', action='store_true', default=False)
         parser.add_argument('--no-input', action='store_true', default=False)
         parser.add_argument('--debug', action='store_true', default=False)
+        parser.add_argument('--run_timestamp', action='store_true', default=None)
```
Comment (millerdev) [inline corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py]: Is it necessary to specify this option to make the migration resumable? If yes, why not enable it by default and (possibly) allow opt-out with an option like `--no-resume`? That suggests the question, is there any reason one would not want to do a resumable migration?

Maybe I'm not understanding the purpose of this argument? A `help="explain what this option does..."` argument would be nice here.

nit: `--run-timestamp` (dash, not underscore) would be more consistent with other options.

File: corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py [hunk 2]
```diff
@@ -59,11 +59,30 @@ def data_function(**view_kwargs):
...
+
+    @staticmethod
+    def _get_paginated_iterable(data_function, args_provider, event_handler=None, resumable_key=None):
+        if resumable_key:
+            return ResumableFunctionIterator(
+                resumable_key,
+                data_function,
+                args_provider,
+                lambda x: x.id,
+                event_handler=event_handler)
```
Comment (millerdev) [inline corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py]: cool! that was easy

nit: paren should be wrapped to new line

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 3]
```diff
@@ -160,6 +161,30 @@ def _process_main_forms(self):
         while not pool.join(timeout=10):
             self.log_info('Waiting on {} docs'.format(len(pool)))
 
+    def _try_to_process_form(self, wrapped_form, pool):
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Yeah, I guess I only put the comment why into the commit message: `putting all forms through queues so they get stored persistently`. Basically, I want all processing forms to be at least in the persistent part of the queue

[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: a43ea98

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 4]
```diff
@@ -160,6 +161,30 @@ def _process_main_forms(self):
...
+        while True:
+            new_wrapped_form = self.queues.get_next()
+            if not new_wrapped_form:
+                break
+            pool.spawn(self._migrate_form_and_associated_models_async, new_wrapped_form)
+
+    def _rebuild_queues(self, pool):
+        prev_ids = self.queues.get_ids_from_run_timestamp()
+        form_db = dbaccessors.FormAccessors(self.domain)
+        for form_id in prev_ids:
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: efcfd07

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 5]
```diff
@@ -742,6 +776,22 @@ def get_queue_obj_id(queue_obj):
...
+            return "partial_queues.queued_or_processing.%s" % run_timestamp
+        self.get_cached_list_key = get_cached_list_key
+
+    def add_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        client.rpush(self.get_cached_list_key(), doc_id)
+
+    def remove_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        client.lrem(self.get_cached_list_key(), 1, doc_id)
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: I would be surprised if it got to be too long i was operating under the assumption that if it were longer than [10000](https://github.com/dimagi/commcare-hq/blob/efcfd074e1ec404c7654f45e2166462e96b374bb/corehq/apps/couch_sql_migration/couchsqlmigration.py#L127), we should [slow down growth](https://github.com/dimagi/commcare-hq/blob/efcfd074e1ec404c7654f45e2166462e96b374bb/corehq/apps/couch_sql_migration/couchsqlmigration.py#L168) and try to process as much from it as possible. i'd be surprised if it ever got there tho. 

i could spin it off into a greenlet since this could be a bottleneck and, since redis is single threaded before v4 (which i don't think we're on), and is supposed to be thread safe, it wouldn't be an issue

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 6]
```diff
@@ -742,6 +776,22 @@ def get_queue_obj_id(queue_obj):
...
+
+    def get_ids_from_run_timestamp(self):
+        client = get_redis_connection()
+        return client.get(self.get_cached_list_key()) or []
```
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: 19aec5e

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: 19aec5e

File: corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py (hunk 1 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py]: Nope, I should definitely explain this better:

It'll always be resumable. This option makes it resume to the run started at the specified run_timestamp. 

So if a previous run happened at `1529941054`, that would have been printed at the beginning of the run and the resumable pieces of that run would be keyed on that timestamp, to be picked up whenever.

File: corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py (hunk 2 above)
Comment (jmtroth0) [reply corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py]: haha, i'm actually not 100% sure which style i prefer

File: corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py (hunk 1 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/management/commands/migrate_domain_from_couch_to_sql.py]: 313d238

File: corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py (hunk 2 above)
Comment (jmtroth0) [reply corehq/ex-submodules/pillowtop/reindexer/change_providers/couch.py]: 732c3b9, there are definitely times that i think it looks nicer with the paren not wrapped

**Comment by jmtroth0** [comment]
@snopoke @millerdev updated for/responded to all comments

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 7]
```diff
@@ -753,6 +802,8 @@ def try_obj(self, lock_ids, queue_obj):
 
         Returns :boolean: True if it acquired the lock, False if it was added to queue
         """
+        if not lock_ids:
+            self._add_item(lock_ids, queue_obj, to_queue=False)
```
Comment (snopoke) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: should this `return True` afterwards?

Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Ah yeah. 001562d

Reading through the code I'm realizing that tho this short circuits the process and arguably adds some clarity about the no lock_ids case, it isn't necessary and just adds some difficulty when reading it. I'm leaning towards keeping it in, but general thoughts on whether a block like that is good practice in general?

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 8]
```diff
@@ -742,6 +775,22 @@ def get_queue_obj_id(queue_obj):
...
+            return "partial_queues.queued_or_processing.%s" % run_timestamp
+        self.get_cached_list_key = get_cached_list_key
+
+    def add_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        client.rpush(self.get_cached_list_key(), doc_id)
+
+    def remove_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        gevent.spawn(client.lrem, self.get_cached_list_key(), 1, doc_id)
```
Comment (millerdev) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: From https://github.com/dimagi/commcare-hq/pull/21007#discussion_r197841021

> i could spin it off into a greenlet since this could be a bottleneck and, since redis is single threaded before v4 (which i don't think we're on), and is supposed to be thread safe, it wouldn't be an issue

Doing this in a greenlet means that other processing can continue, possibly allowing more items to be added to the queue before `lrem` in this greenlet gets invoked. However, since `lrem` is atomic it _will_ block other mutations of that key at some point, which means that this could still be a bottleneck. We just have less insight into when it will happen because the order of operations now depends on the gevent task scheduler. Do you think there is enough other processing not involving this redis key that this will benefit from concurrency? 

Aside: you may already know this, but it is not trivial to implement a durable task queue with redis (see this [SO question](https://stackoverflow.com/q/27986649/10840) and the [article in the first comment](http://blog.bronto.com/engineering/reliable-queueing-in-redis-part-1/))—it takes more than one list to do it properly. Is that what we're trying to accomplish here?

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 9]
```diff
@@ -742,6 +775,22 @@ def get_queue_obj_id(queue_obj):
             return getattr(queue_obj, queue_id_param)
         self.get_queue_obj_id = get_queue_obj_id
 
+        def get_cached_list_key():
+            return "partial_queues.queued_or_processing.%s" % run_timestamp
+        self.get_cached_list_key = get_cached_list_key
+
+    def add_processing_doc_id(self, doc_id):
+        client = get_redis_connection()
+        client.rpush(self.get_cached_list_key(), doc_id)
```
Comment (millerdev) [inline corehq/apps/couch_sql_migration/couchsqlmigration.py]: Does this key ever expire? If not, will it ever be removed if a migration crashes?

**Review by millerdev — APPROVED** [review]

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 8 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Yeah, there's a decent bit of other processing time because of the numerous requests to couch and pg.

A few notes on my thinking about the redis operation: 
(1) Redis tends to be pretty snappy even for O(n) ops (I did some benchmarking if you're interested, but summary: for a list with 100000 objects, trying to remove something that doesn't exist (the worst case), takes <=4ms always and mostly ~1ms, so network latency might be the main blocker anyway). 
(2) n is rarely going to be that large so this would not be a bottleneck in the vast majority of circumstances. 
(3) And because this isn't an academic question, I'm going to make the argument that even where it is large, this will be removing objects from the front of the list primarily, so worst case scenario analysis isn't the most necessary.
Also, because the other mutations are also I/O operations, it won't block the entire process and I think the pool should be big enough to handle some blocking. 
Last thing was that this is the final operation of the greenlets in the pool, so I think it's better to open up space for a new greenlet rather than block here.
So basically, at the very least it isn't worse to have it scheduled by gevent.

Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: To the aside --> No. This is meant to be a best guess that has _at least_ all the ids that need to be processed. If it reprocesses one, it will error out in pg and continue on rather than saving, with no harm done to the process except an extra ~0.1 seconds.

However, that article was an interesting read, at the least because it describes something not too far off from what I implemented without redis

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 9 above)
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Oh I was thinking there was a default ttl. I'll set a ttl on it for 3 days, since that should save it over a weekend if need be

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 8 above)
Comment (millerdev) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Excellent! Sounds like you've thought through this pretty thoroughly. Thanks for the detailed responses to my pessimistic questions.

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 9 above)
Comment (emord) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: Does this rely on the prod redis server? I think @calellowitz work shows that we think we have a very high eviction rate of redis so I'd be hesitant to rely on anything definitely being there

Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: I guess it would. Are there likely to be followup efforts to fix that? It seems bad that one of our data stores wouldn't store data

Comment (emord) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: There are.

> It seems bad that one of our data stores wouldn't store data

I don't think we ever rely on redis absolutely having the data stored, so I wouldn't count it as a "data store", more of a caching layer that should have a real persistent backup method

File: corehq/apps/couch_sql_migration/couchsqlmigration.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (jmtroth0) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: It turns out that it's on the sql overridden thread at this point, so db accessors directed it to the sql db. I switched this to use FormAccessorCouch, which doesn't have an iter_forms. I don't think it should matter speed-wise though, since there should never be _too_ many things in the queue.

**Comment by jmtroth0** [comment]
@snopoke @millerdev This should be good to go now

**Review by snopoke — APPROVED** [review]

File: corehq/apps/couch_sql_migration/couchsqlmigration.py [hunk 10]
```diff
@@ -739,6 +775,25 @@ def get_queue_obj_id(queue_obj):
             return getattr(queue_obj, queue_id_param)
         self.get_queue_obj_id = get_queue_obj_id
 
+        def get_cached_list_key():
+            return "partial_queues.queued_or_processing.%s" % run_timestamp
+        self.get_cached_list_key = get_cached_list_key
+        client = get_redis_connection()
+        client.rpush(get_cached_list_key(), '')  # hack to make key exist
+        client.expire(get_cached_list_key(), 60 * 60 * 24 * 30)  # 30 days
```
[the remark this answers is not in the document]
Comment (millerdev) [reply corehq/apps/couch_sql_migration/couchsqlmigration.py]: redis eviction rate also noted in https://github.com/dimagi/commcare-hq/pull/21007#discussion_r198176298, fwiw

**Review by millerdev — APPROVED** [review]
Data needed to resume is still stored in redis, right? I believe anything stored in redis will likely be lost within a few hours. To fix that I think we should either use a different data store or fix the issue with redis.

**Comment by millerdev** [comment]
> Is it really that quick that things are evicted?

I'm not sure if it's a few hours or a few days, but it's certainly been bad enough lately that I would not use redis as a primary data store for anything until the current issue is resolved. I believe @NitigyaS is expecting to look into it sometime in the next two weeks.

> Is there something else that we have that might be able to handle this easily

Redis does seem like the ideal way to handle this type of thing. Couch might also be an easy solution. Since the migration is done by a single process I think you could keep the state in memory and simply write a snapshot to the DB periodically. It could even be written to Postgres, although it's likely to be more work to setup if we don't already have a table to store random key/value data.

**Comment by jmtroth0** [comment]
>in the next two weeks

If it's in the next 2 weeks, I'd prefer not to overhaul this with a lesser solution. If there were a plug and play option, I'd be happy to swap it out, but I'm not expecting to use this functionality in that time frame.

Do we have a couch setup right now that doesn't go through an orm at all? I've mostly seen the interfaces that pull from particular views. This data really has no need for persistences past a few days, so I wouldn't want to set up anything too permanent for it. And unfortunately occasional snapshots might not work, since keeping the additions up to date is a basic requirement.

**Comment by snopoke** [comment]
@jmtroth0 you don't have to use a ORM object - you can just use a dict or you can use a simple JsonObject to make it easier to use in python a then serialize to a dict when you save it. As an example you can look here:

* Simple model: https://github.com/dimagi/commcare-hq/blob/c64a6187c550da382eff417c7b8f7a7a5e63eb39/corehq/util/pagination.py#L136
* Get's persisted here: https://github.com/dimagi/commcare-hq/blob/c64a6187c550da382eff417c7b8f7a7a5e63eb39/corehq/util/pagination.py#L279

**Comment by millerdev** [comment]
>>> It seems like this would be a pretty slow operation for most dbs, since I'd have to read then write.
>
> And unfortunately occasional snapshots might not work, since keeping the additions up to date is a basic requirement.

I must be missing something. It's my understanding that the current implementation follows a pattern similar to

- `add_processing_doc_id`: push doc ID into list in redis
- `remove_processing_doc_id`: remove first occurrence of doc ID from left end of list in redis
- `get_ids_from_run_timestamp`: get all ids from list in redis

I think you can maintain that list in python and occasionally save it out to disk, as in:

```py
def __init__(self, ...):
    self.key = ...
    self.cached_list = couch.get(self.key) or []  # pseudo code

def add_processing_doc_id(self, doc_id):
    self.cached_list.append(doc_id)
    couch.put(self.key, self.cached_list)  # pseudo code

def remove_processing_doc_id(self, doc_id):
    try:
        self.cached_list.remove(doc_id)
    except ValueError:
        pass
    else:
        # Maybe only do this periodically if high frequency
        # and not critical for data integrity. Reprocessing doc
        # ids on resume is not a problem, right?
        couch.put(self.key, self.cached_list)  # pseudo code

def get_ids_from_run_timestamp(self):
    return self.cached_list
```

To summarize, this solution uses only write operations except for the initial read when the process starts.

**Comment by jmtroth0** [comment]
@millerdev Those two comments were at different points in the thread and refer to different implementations. The read/write comment was before you suggested keeping the primary source of truth in memory, which isn't hard to implement.

The other thing with this implementation that I mentioned is the fact that the removals are happening in greenlets, which means there will have to be some kind of locking around the couch puts or there is the potential for couch accepting an out of date list, which isn't something I'd have to worry about with redis handling the additions and removals.

Again, this isn't difficult to implement, it's just an annoying addition that I would rather not deal with if redis is going to be fixed in a week or two

**Comment by millerdev** [comment]
It's fine if you want to wait, although there is the potential that it would take longer than two weeks to fix redis.

> removals are happening in greenlets, which means there will have to be some kind of locking around the couch puts or there is the potential for couch accepting an out of date list

The list is maintained in Python memory, so whatever version is written at any time will always be up to date. The write to couch on removal could even be skipped entirely if new additions are frequent enough and the wind down at the end (after the last addition) is not very long.

**Comment by jmtroth0** [comment]
As I said, this isn't insurmountable and finding solutions isn't difficult, but it does slow things down a decent amount to do this robustly in pretty much anything else
